### PR TITLE
[midnight adapter] eager loss realization

### PR DESCRIPTION
**Pull request by adhusson**
(ignore the branch name)

### CODE REVIEW

File: src/adapters/MidnightAdapter.sol [hunk 1]
```diff
@@ -473,71 +386,70 @@ contract MidnightAdapter is IMidnightAdapter {
...
-            .deallocate(address(this), abi.encode(ids(market), -int256(netCreditDecrease)), sellerAssets);
-
-        uint256 loss = adapterTotalAssetsBefore.zeroFloorSub(uint256(totalAssets) + sellerAssets);
-        uint256 consumed = loss.zeroFloorSub(vaultBuffer);
-        if (consumed > 0) {
-            skipBufferAllowance -= consumed;
-            emit ConsumeSkipBufferAllowance(consumed);
+            vaultRealAssetsAfter += IAdapter(IVaultV2(parentVault).adapters(i)).realAssets();
         }
+        require(vaultRealAssetsAfter >= IVaultV2(parentVault).totalAssets(), BufferTooLow());
```
Comment (adhusson) [inline src/adapters/MidnightAdapter.sol]: I removed the "allowance" check where we measure how much the vault has lost because i don't know how to make it work with nested sells, which can be triggered by the buyer's callback.

File: src/adapters/MidnightAdapter.sol [hunk 2]
```diff
@@ -125,6 +117,7 @@ contract MidnightAdapter is IMidnightAdapter {
     }
 
     function isRatified(Offer memory offer, bytes memory data, address taker) external view returns (bytes32) {
+        require(IVaultV2(parentVault).firstTotalAssets() != 0, VaultNotAccrued());
```
Comment (adhusson) [inline src/adapters/MidnightAdapter.sol:121]: note this is not great at all since you can't take an adapter offer unless the vault was pinged first

File: src/adapters/MidnightAdapter.sol [hunk 3]
```diff
@@ -285,61 +269,24 @@ contract MidnightAdapter is IMidnightAdapter {
...
+            bytes32 marketId = marketIds[i];
+            MarketData memory marketData = _markets[marketId];
+            uint256 newNetCredit;
+            if (marketData.lossFactor == IMidnight(midnight).lossFactor(marketId)) {
+                newNetCredit = marketData.netCredit;
+            } else {
+                // updatePositionView only reads market.maturity.
+                Market memory market;
+                market.maturity = marketData.maturity;
+                newNetCredit = currentNetCredit(marketId, market);
```
Comment (adhusson) [inline src/adapters/MidnightAdapter.sol:292]: saves 13k gas per market compared to

```solidity
newNetCredit = currentNetCredit(marketId, IMidnight(midnight).toMarket(marketId));
```

File: src/adapters/MidnightAdapter.sol (hunk 2 above)
Comment (adhusson) [reply src/adapters/MidnightAdapter.sol]: not anymore

File: src/adapters/interfaces/IMidnightAdapter.sol [hunk 4]
```diff
@@ -7,15 +7,18 @@ import {Market, Offer} from "lib/midnight/src/interfaces/IMidnight.sol";
 import {IBuyCallback, ISellCallback} from "lib/midnight/src/interfaces/ICallbacks.sol";
 import {IRatifier} from "lib/midnight/src/interfaces/IRatifier.sol";
 
-struct MaturityData {
+struct MarketData {
     uint128 netCredit;
-    uint120 growth;
-    uint8 durationCount;
+    uint128 lossFactor;
```
Comment (MathisGD) [inline src/adapters/interfaces/IMidnightAdapter.sol]: not used

File: src/adapters/MidnightAdapter.sol [hunk 5]
```diff
@@ -389,134 +377,131 @@ contract MidnightAdapter is IMidnightAdapter {
...
     }
 
     /* INTERNAL FUNCTIONS */
 
-    function currentNetCredit(bytes32 marketId) internal view returns (uint256) {
-        return
-            IMidnight(midnight).credit(marketId, address(this))
-                - IMidnight(midnight).pendingFee(marketId, address(this));
+    function currentNetCredit(bytes32 marketId, Market memory market) internal view returns (uint128) {
+        (uint128 credit, uint128 pendingFee,) = IMidnight(midnight).updatePositionView(market, marketId, address(this));
```
Comment (MathisGD) [inline src/adapters/MidnightAdapter.sol:456]: you didn't do the skip if lossFactor didn't change optimization on purpose?

File: src/adapters/MidnightAdapter.sol [hunk 6]
```diff
@@ -123,11 +123,11 @@ contract MidnightAdapter is IMidnightAdapter {
     }
 
     function isRatified(Offer memory offer, bytes memory data, address taker) external view returns (bytes32) {
+        require(!IMidnight(midnight).liquidationLocked(IdLib.toId(offer.market), address(this)), SellInProgress());
         // Collaterals will be checked through vault ids.
```
Comment (MathisGD) [inline src/adapters/MidnightAdapter.sol]:
```suggestion
        // Collaterals and duration will be checked through vault ids.
```

File: src/adapters/MidnightAdapter.sol [hunk 7]
```diff
@@ -238,29 +238,22 @@ contract MidnightAdapter is IMidnightAdapter {
...
-
-    function withdrawToVaultInternal(Market memory market, uint256 withdrawnAssets, uint256 pendingBoughtNetCredit)
-        internal
-    {
         bytes32 marketId = IdLib.toId(market);
-        require(!IMidnight(midnight).liquidationLocked(marketId, address(this)), SellInProgress());
+        require(
+            IVaultV2(parentVault).isAllocator(msg.sender) || IVaultV2(parentVault).isSentinel(msg.sender),
+            NotAuthorized()
+        );
```
Comment (MathisGD) [inline src/adapters/MidnightAdapter.sol]: why not permissionless?

File: src/adapters/MidnightAdapter.sol (hunk 5 above)
Comment (adhusson) [reply src/adapters/MidnightAdapter.sol:456]: i used to do it in `realAssets` (in `currentNetCredit` it would be too aggressive sometimes you check after a credit change unrelated to a loss)

```solidity
if (marketId == overridenMarketId) {
    newNetCredit = overridenMarketNetCredit;
if (marketData.lossFactor == IMidnight(midnight).lossFactor(marketId)) {
    newNetCredit = marketData.netCredit;
} else {
    <currentNetCredit>
}
```

but it didn't seem worth the extra branch & field (the one i forgot to remove).

I checked gas diff, for each market its 8k saved on cold, 4k saved on warm, and +1.5k if the loss factor did change (which is rarer). 

wdyt? still not sure it's worth the extra branch.

Comment (MathisGD) [reply src/adapters/MidnightAdapter.sol:456]: indeed it's an extra branch but with a very simple logic, sounds ok to me. we can re-add later too

File: src/adapters/MidnightAdapter.sol (hunk 7 above)
Comment (adhusson) [reply src/adapters/MidnightAdapter.sol]: lost in a merge

File: src/adapters/MidnightAdapter.sol (hunk 5 above)
Comment (MathisGD) [reply src/adapters/MidnightAdapter.sol:456]: let's merge this and re-add later
